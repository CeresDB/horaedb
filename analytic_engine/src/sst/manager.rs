// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.

//! Multi-level SST management

use common_types::time::{TimeRange, Timestamp};

use crate::{
    compaction::ExpiredFiles,
    sst::file::{FileHandle, FileMeta, FilePurgeQueue, Iter, Level, LevelHandler},
};

/// Id for a sst file
pub type FileId = u64;
/// We use two level merge tree, the max level should less than u16::MAX
pub const MAX_LEVEL: usize = 2;

/// A table level manager that manages all the sst files of the table
pub struct LevelsController {
    levels: Vec<LevelHandler>,
    purge_queue: FilePurgeQueue,
}

impl Drop for LevelsController {
    fn drop(&mut self) {
        // Close the purge queue to avoid files being deleted.
        self.purge_queue.close();
    }
}

impl LevelsController {
    /// Create an empty LevelsController
    pub fn new(purge_queue: FilePurgeQueue) -> Self {
        let mut levels = Vec::with_capacity(MAX_LEVEL);
        for level in 0..MAX_LEVEL {
            levels.push(LevelHandler::new(level as Level));
        }

        Self {
            levels,
            purge_queue,
        }
    }

    /// Add sst file to level
    ///
    /// Panic: If the level is greater than the max level
    pub fn add_sst_to_level(&mut self, level: Level, file_meta: FileMeta) {
        let level_handler = &mut self.levels[usize::from(level)];
        let file = FileHandle::new(file_meta, self.purge_queue.clone());

        level_handler.insert(file);
    }

    pub fn latest_sst(&self, level: Level) -> Option<FileHandle> {
        self.levels[usize::from(level)].latest_sst()
    }

    /// Pick the ssts and collect it by `append_sst`.
    pub fn pick_ssts(
        &self,
        time_range: TimeRange,
        mut append_sst: impl FnMut(Level, &[FileHandle]),
    ) {
        for level_handler in self.levels.iter() {
            let ssts = level_handler.pick_ssts(time_range);
            append_sst(level_handler.level, &ssts);
        }
    }

    /// Remove sst files from level.
    ///
    /// Panic: If the level is greater than the max level
    pub fn remove_ssts_from_level(&mut self, level: Level, file_ids: &[FileId]) {
        let level_handler = &mut self.levels[usize::from(level)];
        level_handler.remove_ssts(file_ids);
    }

    /// Total number of levels.
    pub fn num_levels(&self) -> Level {
        self.levels.len() as Level
    }

    /// Iter ssts at given `level`.
    ///
    /// Panic if level is out of bound.
    pub fn iter_ssts_at_level(&self, level: Level) -> Iter {
        let level_handler = &self.levels[usize::from(level)];
        level_handler.iter_ssts()
    }

    pub fn collect_expired_at_level(
        &self,
        level: Level,
        expire_time: Option<Timestamp>,
    ) -> Vec<FileHandle> {
        let level_handler = &self.levels[usize::from(level)];
        let mut expired = Vec::new();
        level_handler.collect_expired(expire_time, &mut expired);

        expired
    }

    pub fn has_expired_sst(&self, expire_time: Option<Timestamp>) -> bool {
        self.levels
            .iter()
            .any(|level_handler| level_handler.has_expired_sst(expire_time))
    }

    pub fn expired_ssts(&self, expire_time: Option<Timestamp>) -> Vec<ExpiredFiles> {
        let mut expired = Vec::new();
        let num_levels = self.num_levels();
        for level in 0..num_levels {
            let files = self.collect_expired_at_level(level, expire_time);
            expired.push(ExpiredFiles { level, files });
        }

        expired
    }
}

#[cfg(test)]
pub mod tests {
    use table_engine::table::TableId;
    use tokio::sync::mpsc;

    use crate::{
        sst::{
            file::{FileMeta, FilePurgeQueue},
            manager::{FileId, LevelsController},
            meta_data::SstMetaData,
        },
        table_options::StorageFormat,
    };

    #[must_use]
    #[derive(Default)]
    pub struct LevelsControllerMockBuilder {
        sst_meta_vec: Vec<SstMetaData>,
    }

    impl LevelsControllerMockBuilder {
        pub fn add_sst(mut self, mut sst_meta: Vec<SstMetaData>) -> Self {
            self.sst_meta_vec.append(&mut sst_meta);
            self
        }

        pub fn build(self) -> LevelsController {
            let (tx, _rx) = mpsc::unbounded_channel();
            let file_purge_queue = FilePurgeQueue::new(100, TableId::from(101), tx);
            let mut levels_controller = LevelsController::new(file_purge_queue);
            for (id, sst_meta) in self.sst_meta_vec.into_iter().enumerate() {
                levels_controller.add_sst_to_level(
                    0,
                    FileMeta {
                        id: id as FileId,
                        size: 0,
                        row_num: 0,
                        time_range: sst_meta.time_range(),
                        max_seq: sst_meta.max_sequence(),
                        storage_format: StorageFormat::Columnar,
                    },
                );
            }
            levels_controller
        }
    }
}
