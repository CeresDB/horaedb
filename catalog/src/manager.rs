// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.

//! Catalog manager

use std::sync::Arc;

use snafu::Snafu;

use crate::{schema::NameRef, CatalogRef};

#[derive(Debug, Snafu)]
pub struct Error;

define_result!(Error);

/// Catalog manager abstraction
///
/// Tracks meta data of databases/tables
// TODO(yingwen): Maybe use async trait?
// TODO(yingwen): Provide a context

pub trait Manager: Send + Sync {
    /// Get the default catalog name
    fn default_catalog_name(&self) -> NameRef;

    /// Get the default schema name
    fn default_schema_name(&self) -> NameRef;

    /// Find the catalog by name
    fn catalog_by_name(&self, name: NameRef) -> Result<Option<CatalogRef>>;

    /// All catalogs
    fn all_catalogs(&self) -> Result<Vec<CatalogRef>>;
}

pub type ManagerRef = Arc<dyn Manager>;
