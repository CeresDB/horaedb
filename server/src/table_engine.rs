// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.

//! Table engine implementation

use std::sync::Arc;

use async_trait::async_trait;
use table_engine::{
    engine::{
        CloseTableRequest, CreateTableRequest, DropTableRequest, OpenTableRequest, Result,
        TableEngine, TableEngineRef, UnknownEngineType,
    },
    memory::MemoryTable,
    table::TableRef,
    ANALYTIC_ENGINE_TYPE, MEMORY_ENGINE_TYPE,
};

/// Memory table engine implementation
// Mainly for test purpose now
pub struct MemoryTableEngine;

#[async_trait]
impl TableEngine for MemoryTableEngine {
    fn engine_type(&self) -> &str {
        MEMORY_ENGINE_TYPE
    }

    async fn close(&self) -> Result<()> {
        Ok(())
    }

    async fn create_table(&self, request: CreateTableRequest) -> Result<TableRef> {
        Ok(Arc::new(MemoryTable::new(
            request.table_name,
            request.table_id,
            request.table_schema,
            MEMORY_ENGINE_TYPE.to_string(),
        )))
    }

    async fn drop_table(&self, _request: DropTableRequest) -> Result<bool> {
        Ok(true)
    }

    async fn open_table(&self, _request: OpenTableRequest) -> Result<Option<TableRef>> {
        Ok(None)
    }

    async fn close_table(&self, _request: CloseTableRequest) -> Result<()> {
        Ok(())
    }
}

/// Route [CreateTableRequest] to the correct engine by its engine type
pub struct TableEngineProxy {
    /// Memory table engine
    pub memory: MemoryTableEngine,
    /// Analytic table engine
    pub analytic: TableEngineRef,
}

#[async_trait]
impl TableEngine for TableEngineProxy {
    fn engine_type(&self) -> &str {
        "TableEngineProxy"
    }

    async fn close(&self) -> Result<()> {
        self.memory.close().await?;
        self.analytic.close().await?;

        Ok(())
    }

    async fn create_table(&self, request: CreateTableRequest) -> Result<TableRef> {
        // TODO(yingwen): Use a map
        match request.engine.as_str() {
            MEMORY_ENGINE_TYPE => self.memory.create_table(request).await,
            ANALYTIC_ENGINE_TYPE => self.analytic.create_table(request).await,
            engine_type => UnknownEngineType { engine_type }.fail(),
        }
    }

    async fn drop_table(&self, request: DropTableRequest) -> Result<bool> {
        match request.engine.as_str() {
            MEMORY_ENGINE_TYPE => self.memory.drop_table(request).await,
            ANALYTIC_ENGINE_TYPE => self.analytic.drop_table(request).await,
            engine_type => UnknownEngineType { engine_type }.fail(),
        }
    }

    /// Open table, return error if table not exists
    async fn open_table(&self, request: OpenTableRequest) -> Result<Option<TableRef>> {
        match request.engine.as_str() {
            MEMORY_ENGINE_TYPE => self.memory.open_table(request).await,
            ANALYTIC_ENGINE_TYPE => self.analytic.open_table(request).await,
            engine_type => UnknownEngineType { engine_type }.fail(),
        }
    }

    /// Close table, it is ok to close a closed table.
    async fn close_table(&self, request: CloseTableRequest) -> Result<()> {
        match request.engine.as_str() {
            MEMORY_ENGINE_TYPE => self.memory.close_table(request).await,
            ANALYTIC_ENGINE_TYPE => self.analytic.close_table(request).await,
            engine_type => UnknownEngineType { engine_type }.fail(),
        }
    }
}
