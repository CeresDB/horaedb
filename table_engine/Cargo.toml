[package]
name = "table_engine"

[package.version]
workspace = true

[package.authors]
workspace = true

[package.edition]
workspace = true

[dependencies]
# In alphabetical order
arrow = { workspace = true }
arrow_ext = { workspace = true }
async-trait = { workspace = true }
ceresdbproto = { workspace = true }
common_types = { workspace = true }
common_util = { workspace = true }
datafusion = { workspace = true }
datafusion-proto = { workspace = true }
df_operator = { workspace = true }
futures = { workspace = true }
itertools = { workspace = true }
log = { workspace = true }
prost = { workspace = true }
serde = { workspace = true }
smallvec = { workspace = true }
snafu = { workspace = true }
tokio = { workspace = true }
trace_metric = { workspace = true }

[dev-dependencies]
env_logger = { workspace = true }
