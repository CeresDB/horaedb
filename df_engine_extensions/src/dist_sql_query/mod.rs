// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::{fmt, sync::Arc};

use async_trait::async_trait;
use datafusion::{
    error::Result as DfResult,
    physical_plan::{ExecutionPlan, SendableRecordBatchStream},
};
use table_engine::{
    remote::model::TableIdentifier,
    table::{ReadRequest, TableRef},
};

pub mod codec;
pub mod physical_plan;
pub mod resolver;
#[cfg(test)]
pub mod test_util;

/// Remote datafusion physical plan executor
#[async_trait]
pub trait RemotePhysicalPlanExecutor: Clone + fmt::Debug + Send + Sync + 'static {
    async fn execute(
        &self,
        table: TableIdentifier,
        physical_plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<SendableRecordBatchStream>;
}

/// Executable scan's builder
///
/// It is not suitable to restrict the detailed implementation of executable
/// scan, so we define a builder here which return the general `ExecutionPlan`.
pub trait ExecutableScanBuilder: fmt::Debug + Send + Sync + 'static {
    fn build(&self, table: TableRef, read_request: ReadRequest)
        -> DfResult<Arc<dyn ExecutionPlan>>;
}
