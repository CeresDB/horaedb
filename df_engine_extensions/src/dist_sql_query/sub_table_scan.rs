// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.

use std::{any::Any, fmt, sync::Arc};

use arrow::datatypes::SchemaRef;
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    execution::TaskContext,
    physical_expr::PhysicalSortExpr,
    physical_plan::{
        DisplayAs, DisplayFormatType, ExecutionPlan, Partitioning,
        SendableRecordBatchStream as DfSendableRecordBatchStream, Statistics,
    },
};
use table_engine::{provider::ScanTable, remote::model::TableIdentifier, table::ReadRequest};

/// Unresolved sub table scan which can't be executed before rewriting
#[derive(Debug)]
pub struct UnresolvedSubTableScan {
    pub table: TableIdentifier,
    pub read_request: ReadRequest,
}

impl ExecutionPlan for UnresolvedSubTableScan {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> SchemaRef {
        self.read_request
            .projected_schema
            .to_projected_arrow_schema()
    }

    fn output_partitioning(&self) -> Partitioning {
        Partitioning::UnknownPartitioning(self.read_request.opts.read_parallelism)
    }

    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
        None
    }

    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
        vec![]
    }

    fn with_new_children(
        self: Arc<Self>,
        _children: Vec<Arc<dyn ExecutionPlan>>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        Err(DataFusionError::Internal(
            "UnresolvedSubTableScan should not have children".to_string(),
        ))
    }

    fn execute(
        &self,
        _partition: usize,
        _context: Arc<TaskContext>,
    ) -> DfResult<DfSendableRecordBatchStream> {
        Err(DataFusionError::Internal(
            "UnresolvedSubTableScan can not be executed".to_string(),
        ))
    }

    fn statistics(&self) -> Statistics {
        Statistics::default()
    }
}

impl DisplayAs for UnresolvedSubTableScan {
    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "UnresolvedSubTableScan: table={:?}, read_request:{:?}, partition_count={}",
            self.table,
            self.read_request,
            self.output_partitioning().partition_count(),
        )
    }
}

/// `ResolvedSubTableScan` is `ScanTable` actually.
pub type ResolvedSubTableScan = ScanTable;
