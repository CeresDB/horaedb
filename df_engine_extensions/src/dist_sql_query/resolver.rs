// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use catalog::manager::ManagerRef as CatalogManagerRef;
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    physical_plan::ExecutionPlan,
};
use table_engine::{remote::model::TableIdentifier, table::TableRef};

use crate::dist_sql_query::{
    physical_plan::{ResolvedPartitionedScan, UnresolvedPartitionedScan, UnresolvedSubTableScan},
    ExecutableScanBuilder, RemotePhysicalPlanExecutor,
};

/// Resolver which makes datafuison dist query related plan executable.
///
/// The reason we define a `Resolver` rather than `physical optimization rule`
/// is: As I see, physical optimization rule is responsible for optimizing a bad
/// plan to good one, rather than making a inexecutable plan executable.
/// So we define `Resolver` to make it, it may be somthing similar to task
/// generator responsible for generating task for executor to run based on
/// physical plan.
pub trait Resolver {
    fn resolve(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>>;
}

/// Resolver which makes the partitioned table scan plan executable
struct PartitionedScanResolver<R> {
    remote_executor: R,
}

impl<R: RemotePhysicalPlanExecutor> PartitionedScanResolver<R> {
    #[allow(dead_code)]
    pub fn new(remote_executor: R) -> Self {
        Self { remote_executor }
    }

    fn resolve_plan(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedPartitionedScan>() {
            let sub_tables = unresolved.sub_tables.clone();
            let remote_plans = sub_tables
                .into_iter()
                .map(|table| {
                    let plan = Arc::new(UnresolvedSubTableScan {
                        table: table.clone(),
                        read_request: unresolved.read_request.clone(),
                    });
                    (table, plan as _)
                })
                .collect::<Vec<_>>();

            return Ok(Arc::new(ResolvedPartitionedScan {
                remote_executor: self.remote_executor.clone(),
                remote_exec_plans: remote_plans,
            }));
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_plan(child)?;

            new_children.push(child);
        }

        // TODO: Push down the computation physical node here rather than simply
        // rebuild.
        plan.with_new_children(new_children)
    }
}

impl<R: RemotePhysicalPlanExecutor> Resolver for PartitionedScanResolver<R> {
    fn resolve(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>> {
        self.resolve_plan(plan)
    }
}

/// Resolver which makes the sub table scan plan executable
struct SubScanResolver<B> {
    catalog_manager: CatalogManagerRef,
    scan_builder: B,
}

impl<B: ExecutableScanBuilder> SubScanResolver<B> {
    #[allow(dead_code)]
    pub fn new(catalog_manager: CatalogManagerRef, scan_builder: B) -> Self {
        Self {
            catalog_manager,
            scan_builder,
        }
    }

    fn resolve_plan(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedSubTableScan>() {
            let table = self.find_table(&unresolved.table)?;
            return self
                .scan_builder
                .build(table, unresolved.read_request.clone());
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_plan(child)?;

            new_children.push(child);
        }

        plan.with_new_children(new_children)
    }

    fn find_table(&self, table_ident: &TableIdentifier) -> DfResult<TableRef> {
        let catalog = self
            .catalog_manager
            .catalog_by_name(&table_ident.catalog)
            .map_err(|e| DataFusionError::Internal(format!("failed to find catalog, err:{e}")))?
            .ok_or(DataFusionError::Internal("catalog not found".to_string()))?;

        let schema = catalog
            .schema_by_name(&table_ident.schema)
            .map_err(|e| DataFusionError::Internal(format!("failed to find schema, err:{e}")))?
            .ok_or(DataFusionError::Internal("schema not found".to_string()))?;

        schema
            .table_by_name(&table_ident.table)
            .map_err(|e| DataFusionError::Internal(format!("failed to find table, err:{e}")))?
            .ok_or(DataFusionError::Internal("table not found".to_string()))
    }
}

impl<B: ExecutableScanBuilder> Resolver for SubScanResolver<B> {
    fn resolve(&self, plan: Arc<dyn ExecutionPlan>) -> DfResult<Arc<dyn ExecutionPlan>> {
        self.resolve_plan(plan)
    }
}

#[cfg(test)]
mod test {
    use std::sync::Arc;

    use async_trait::async_trait;
    use datafusion::{
        error::Result as DfResult,
        physical_plan::{displayable, DisplayAs, ExecutionPlan, SendableRecordBatchStream},
    };
    use table_engine::{
        remote::model::TableIdentifier,
        table::{ReadRequest, TableRef},
    };

    use crate::dist_sql_query::{
        resolver::{PartitionedScanResolver, RemotePhysicalPlanExecutor, SubScanResolver},
        test_util::TestContext,
        ExecutableScanBuilder,
    };

    #[test]
    fn test_resolve_simple_partitioned_scan() {
        let ctx = TestContext::new();
        let plan = ctx.build_basic_partitioned_table_plan();
        let resolver = PartitionedScanResolver {
            remote_executor: MockRemotePhysicalPlanExecutor,
        };
        let new_plan = displayable(resolver.resolve_plan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[test]
    fn test_resolve_simple_sub_scan() {
        let ctx = TestContext::new();
        let plan = ctx.build_basic_sub_table_plan();
        let resolver = SubScanResolver {
            catalog_manager: ctx.catalog_manager(),
            scan_builder: MockScanBuilder,
        };
        let new_plan = displayable(resolver.resolve_plan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    // Mock scan and its builder
    #[derive(Debug)]
    struct MockScanBuilder;

    impl ExecutableScanBuilder for MockScanBuilder {
        fn build(
            &self,
            _table: TableRef,
            read_request: ReadRequest,
        ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
            Ok(Arc::new(MockScan {
                request: read_request,
            }))
        }
    }

    #[derive(Debug)]
    struct MockScan {
        request: ReadRequest,
    }

    impl ExecutionPlan for MockScan {
        fn as_any(&self) -> &dyn std::any::Any {
            unimplemented!()
        }

        fn schema(&self) -> arrow::datatypes::SchemaRef {
            self.request.projected_schema.to_projected_arrow_schema()
        }

        fn output_partitioning(&self) -> datafusion::physical_plan::Partitioning {
            datafusion::physical_plan::Partitioning::UnknownPartitioning(
                self.request.opts.read_parallelism,
            )
        }

        fn output_ordering(&self) -> Option<&[datafusion::physical_expr::PhysicalSortExpr]> {
            None
        }

        fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
            vec![]
        }

        fn with_new_children(
            self: Arc<Self>,
            _children: Vec<Arc<dyn ExecutionPlan>>,
        ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
            unimplemented!()
        }

        fn execute(
            &self,
            _partition: usize,
            _context: Arc<datafusion::execution::TaskContext>,
        ) -> datafusion::error::Result<datafusion::physical_plan::SendableRecordBatchStream>
        {
            unimplemented!()
        }

        fn statistics(&self) -> datafusion::physical_plan::Statistics {
            unimplemented!()
        }
    }

    impl DisplayAs for MockScan {
        fn fmt_as(
            &self,
            _t: datafusion::physical_plan::DisplayFormatType,
            f: &mut std::fmt::Formatter,
        ) -> std::fmt::Result {
            write!(f, "MockScan")
        }
    }

    // Mock remote executor
    #[derive(Debug, Clone)]
    struct MockRemotePhysicalPlanExecutor;

    #[async_trait]
    impl RemotePhysicalPlanExecutor for MockRemotePhysicalPlanExecutor {
        async fn execute(
            &self,
            _table: TableIdentifier,
            _physical_plan: Arc<dyn ExecutionPlan>,
        ) -> DfResult<SendableRecordBatchStream> {
            unimplemented!()
        }
    }
}
