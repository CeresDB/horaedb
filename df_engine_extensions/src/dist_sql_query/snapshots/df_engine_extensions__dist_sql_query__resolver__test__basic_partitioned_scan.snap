---
source: df_engine_extensions/src/dist_sql_query/resolver.rs
assertion_line: 266
expression: new_plan
---
ResolvedPartitionedScan: pushdown_continue:false, partition_count:3
  ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
    FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
      UnresolvedSubTableScan: table:TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_1" }, table_scan_ctx:TableScanContext { read_parallelism: 8, batch_size: 10000, projected: "[time,tag1,tag2,value,field2]", predicate: "[time < TimestampMillisecond(1691974518000, None) AND tag1 = Utf8(\"test_tag\")]" }, partition_count:8
  ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
    FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
      UnresolvedSubTableScan: table:TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_2" }, table_scan_ctx:TableScanContext { read_parallelism: 8, batch_size: 10000, projected: "[time,tag1,tag2,value,field2]", predicate: "[time < TimestampMillisecond(1691974518000, None) AND tag1 = Utf8(\"test_tag\")]" }, partition_count:8
  ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
    FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
      UnresolvedSubTableScan: table:TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_3" }, table_scan_ctx:TableScanContext { read_parallelism: 8, batch_size: 10000, projected: "[time,tag1,tag2,value,field2]", predicate: "[time < TimestampMillisecond(1691974518000, None) AND tag1 = Utf8(\"test_tag\")]" }, partition_count:8

